custom/rch/forms.py

```diff
@@ -0,0 +1,111 @@
+from django.core.urlresolvers import reverse
+from django import forms
+from django.utils.translation import ugettext_lazy as _
+from django.db.utils import ProgrammingError
+from django.conf import settings
+from crispy_forms import layout as crispy
+from crispy_forms.layout import Layout
+from crispy_forms.helper import FormHelper
+from crispy_forms.bootstrap import StrictButton
+from corehq.apps.hqwebapp import crispy as hqcrispy
+from custom.rch.models import AreaMapping
+
+
+def get_choices_for(value_field, display_field):
+    """
+    return a tuple of tuples with value and display text for each unique available option for
+    display field
+    for ex: return all combinations for state code and state name for all unique state names
+    available
+    :param value_field: column for value
+    :param display_field: column for display text
+    :return: tuple of tuples
+    """
+    if settings.UNIT_TESTING:
+        return tuple()
+    options = set()
+    try:
+        possible_display_values = AreaMapping.objects.values_list(display_field, flat=True).distinct()
+        for display_value in possible_display_values:
+            option_values = (
+                AreaMapping.objects.filter(**{display_field: display_value})
+                .values_list(value_field, flat=True).distinct()
+            )
+            for option_value in option_values:
+                options.add((option_value, display_value))
+        return tuple(options)
+    # when migrations are run for the first time to add AreaMapping table this would
+    # fail with Programming Error when loading environment for migration so just pass
+    # then. When the web server/worker would be started migrations would have run already
+    # and the proper values would be set
+    except ProgrammingError:
+        return tuple()
+
+
+class BeneficiariesFilterForm(forms.Form):
+    stcode = forms.ChoiceField(
+        label=_("State"),
+        required=False,
+        choices=get_choices_for('stcode', 'stname'),
```

Comment:
Rather than doing this statically you should do it in the form's `__init__`: https://stackoverflow.com/a/3420588/632517

That will also remove the need to catch `ProgrammingError` in the function

Reply:
Doing it in `__init__` would also mean you could get all the options from one query rather than multiple queries per field

Reply:
Doing this in `__init__` would mean that those queries would run each time the form is rendered. Should not we just call it once since it won't change anyway?

> Doing it in __init__ would also mean you could get all the options from one query rather than multiple queries per field

How is that?

Reply:
> You could cache the query result to avoid running the query every time. This would also mean that you could update the values without having to restart the service.

These values are not expected to change very often. It would be after months it would change so there is no need to worry about the need to refresh it often.

I still think loading them just once on load is still a better call instead of moving to `__init__`. The programming error will just happen once during first migration and after that it would just return results(blank in case of local environment). I would rather have 4 queries fired once on load then have them being called each time on form load or checking for cache in case we cache it.
If you still think it would be better, happy to move it to `__init__`

Reply:
Having it this way means a db query on import, which I think means that that can happen whenever we run `./manage.py anything`.  If you move it to the `__init__` method, you can slap a `@quickcache` decorator on it to cache the results across all workers, or if you prefer, a `@memoized` decorator, which will store it in memory similar to how it's done now.  In either case, it'd be lazily evaluated, so it wouldn't impact startup times.